Quality of the paired end 75â€‰bp reads was assessed using FastQC and reads were then aligned to the {{ snakemake.config["methods"]["genome_name"] }} reference genome using Bowtie2 with the parameters --very-sensitive -X {{ snakemake.config["bowtie2"]["maxins"] }} --no-mixed --no-discordant. PCR duplicate reads were removed using Picard MarkDuplicates. The reads were filtered using samtools for a minimum mapping quality of {{ snakemake.config["bamfilter"]["mapqual"] }} and only properly paired reads were retained. Additionally, any reads aligning to the mitochondrial chromosome were removed. Peaks were called subsequently using macs2 with the parameters --keep-dup all -f BAMPE -q {{ snakemake.config["macs2"]["gsize"] }}, while simultaneously generating bedgraph files in Reads Per Million scale. Peaks were filtered against the ENCODE blacklist regions. The bedgraph files were converted to bigWig file format using the bedGraphToBigWig utility. The quality was acessed using ataqv.
